src/vm/object.cpp

```diff
@@ -1969,54 +1969,160 @@ StackTraceElement & StackTraceArray::operator[](size_t index)
 }
 
 #if !defined(DACCESS_COMPILE)
-// Define the lock used to access stacktrace from an exception object
+// Define the lock used to access stacktrace from a global exception type
 SpinLock g_StackTraceArrayLock;
 
 void ExceptionObject::SetStackTrace(StackTraceArray const & stackTrace, PTRARRAYREF dynamicMethodArray)
 {        
     CONTRACTL
     {
-        GC_NOTRIGGER;
-        NOTHROW;
+        THROWS;
+        GC_TRIGGERS;
         MODE_COOPERATIVE;
     }
     CONTRACTL_END;
 
-    Thread *m_pThread = GetThread();
-    SpinLock::AcquireLock(&g_StackTraceArrayLock, SPINLOCK_THREAD_PARAM_ONLY_IN_SOME_BUILDS);
+    if (GetMethodTable() == g_pOutOfMemoryExceptionClass)
+    {
+        // We can't activate the SyncBlock for an OutOfMemoryException, so use preallocated SpinLock
+        SpinLock::AcquireLock(&g_StackTraceArrayLock, SPINLOCK_THREAD_PARAM_ONLY_IN_SOME_BUILDS);
 
-    SetObjectReference((OBJECTREF*)&_stackTrace, (OBJECTREF)stackTrace.Get());
-    SetObjectReference((OBJECTREF*)&_dynamicMethods, (OBJECTREF)dynamicMethodArray);
+        SetObjectReference((OBJECTREF*)&_stackTrace, (OBJECTREF)stackTrace.Get());
+        SetObjectReference((OBJECTREF*)&_dynamicMethods, (OBJECTREF)dynamicMethodArray);
 
-    SpinLock::ReleaseLock(&g_StackTraceArrayLock, SPINLOCK_THREAD_PARAM_ONLY_IN_SOME_BUILDS);
+        SpinLock::ReleaseLock(&g_StackTraceArrayLock, SPINLOCK_THREAD_PARAM_ONLY_IN_SOME_BUILDS);
+    }
+    else
+    {
+        struct _gc
+        {
+            EXCEPTIONREF throwable;
+            OBJECTREF dynamicMethodArray;
+            OBJECTREF stateLock;
+            OBJECTREF trace;
+        };
+        _gc gc;
+        gc.throwable = (EXCEPTIONREF)this;
+        gc.dynamicMethodArray = (OBJECTREF)dynamicMethodArray;
+        gc.stateLock = _stackTraceLock;
+        gc.trace = (OBJECTREF)stackTrace.Get();
+        GCPROTECT_BEGIN(gc);
+
+        // Aquire lock
+        gc.stateLock->EnterObjMonitor();
```

Comment:
I am worried that there may be cases where the exception object is not going to be fully initialized and we will crash here. You can get an exception like that using `RuntimeHelpers.GetUninitializedObject`.

I would change the condition above to take the global lock path also when `_stackTraceLock` is null.

Reply:
Added